File: 01.Python/0986.py
from typing import List

class Solution:
    def intervalIntersection(self, firstList: List[List[int]], secondList: List[List[int]]) -> List[List[int]]:
        result = []

        i = 0
        j = 0

        while i < len(firstList) and j < len(secondList):
            start1 = firstList[i][0]
            end1 = firstList[i][1]
            
            start2 = secondList[j][0]
            end2 = secondList[j][1]

            if max(start1, start2) <= min(end1, end2):
                result.append([max(start1, start2), min(end1, end2)])

            if end1 <= end2:
                i += 1
            else:
                j += 1

        return result

sol = Solution()

firstList = [[0, 2], [5, 10], [13, 23], [24, 25]]
secondList = [[1, 5], [8, 12], [15, 24], [25, 26]]

result = sol.intervalIntersection(firstList, secondList)
print(result)

File: 03.CPP/0986.cpp
#include <iostream>
#include <vector>

class Solution {
public:
    std::vector<std::vector<int>> intervalIntersection(std::vector<std::vector<int>>& firstList, std::vector<std::vector<int>>& secondList) {
        std::vector<std::vector<int>> result;

        int i = 0;
        int j = 0;

        while ((i < firstList.size()) && (j < secondList.size())) {
            int start1 = firstList[i][0];
            int end1 = firstList[i][1];
            
            int start2 = secondList[j][0];
            int end2 = secondList[j][1];

            if ((std::max(start1, start2)) <= (std::min(end1, end2))) {
                result.push_back({std::max(start1, start2), std::min(end1, end2)});
            }

            if (end1 <= end2) {
                i++;
            } else {
                j++;
            }
        }

        return result;
    }
};

void printVector(std::vector<std::vector<int>>& vec) {
    std::cout << "[";
    for (int i = 0; i < vec.size(); i++) {
        std::cout << "[";
        std::cout << vec[i][0] << ", " << vec[i][1];
        std::cout << "]";

        if (i < vec.size() - 1) {
            std::cout << ", ";
        }
    }
    std::cout << "]" << "\n";
}

int main() {
    Solution sol;

    std::vector<std::vector<int>> firstList = {{0, 2}, {5, 10}, {13, 23}, {24, 25}};
    std::vector<std::vector<int>> secondList = {{1, 5}, {8, 12}, {15, 24}, {25, 26}};

    std::vector<std::vector<int>> result = sol.intervalIntersection(firstList, secondList);

    printVector(result);
    return 0;
}